stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1865,239 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
+      }
+    }
+
+    @inlinable
+    public func tangentVector(
+      from cotangent: Array.CotangentVector
+    ) -> Array.TangentVector {
+      precondition(self.elements.count >= cotangent.elements.count)
+      return Array.TangentVector(elements: self.elements.indices.map {
+        self.elements[$0].tangentVector(
+          from: $0 < cotangent.elements.count ? cotangent.elements[$0] : .zero)
+      })
+    }
+  }
+
+  public struct CotangentVector : AdditiveArithmetic & Differentiable {
+    public typealias AllDifferentiableVariables = Array.CotangentVector
+    public typealias TangentVector = Array.CotangentVector
+    public typealias CotangentVector = Array.TangentVector
+
+    public var elements: [Element.CotangentVector]
+
+    public static var zero: Array.CotangentVector {
+      return Array.CotangentVector(elements: [])
+    }
+    
+    public init(elements: [Element.CotangentVector]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public static func + (
+      lhs: Array.CotangentVector,
+      rhs: Array.CotangentVector
+    ) -> Array.CotangentVector {
+      return Array.CotangentVector(
+        elements: zipLongest(lhs.elements, rhs.elements, default: .zero).map(+))
+    }
+
+    @inlinable
+    public static func - (
+      lhs: Array.CotangentVector,
+      rhs: Array.CotangentVector
+    ) -> Array.CotangentVector {
+      return Array.CotangentVector(
+        elements: zipLongest(lhs.elements, rhs.elements, default: .zero).map(-))
+    }
+
+    @inlinable
+    public func moved(
+      along direction: Array.CotangentVector
+    ) -> [Element.CotangentVector] {
+      precondition(self.elements.count >= direction.elements.count)
+      return elements.indices.map {
+        self.elements[$0].moved(
+          along: $0 < direction.elements.count ? direction.elements[$0] : .zero)
+      }
+    }
+
+    @inlinable
+    public func tangentVector(
+      from cotangent: Array.TangentVector
+    ) -> Array.CotangentVector {
+      precondition(self.elements.count >= cotangent.elements.count)
+      return Array.CotangentVector(elements: self.elements.indices.map {
+        self.elements[$0].tangentVector(
+          from: $0 < cotangent.elements.count ? cotangent.elements[$0] : .zero)
+      })
+    }
+  }
+
+  @inlinable
+  public func moved(along direction: TangentVector) -> [Element] {
+    precondition(self.count >= direction.elements.count)
+    return indices.map {
+      self[$0].moved(
+        along: $0 < direction.elements.count ? direction.elements[$0] : .zero)
+    }
+  }
+
+  @inlinable
+  public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+    precondition(self.count >= cotangent.elements.count)
+    return TangentVector(elements: indices.map {
+      self[$0].tangentVector(
+        from: $0 < cotangent.elements.count ? cotangent.elements[$0] : .zero)
+    })
+  }
+}
+
+@usableFromInline
+internal func zipLongest<S: Sequence>(
```

Comment:
The standard library convention is to underscore internal things. Also, since `@usableFromInline` functions cannot be `fileprivate`, you might want to make the name more specific to its exact use case so it won't collide with other stdlib stuff in the future.

This is not at all the standard Swift style, so hopefully this won't mislead you...

```suggestion
internal func _lazyZipLongestForArrayTangent<S: Sequence>(
```

Reply:
That makes sense. I actually tried to make it `fileprivate` and it was annoying that I couldn't because I felt it is polluting the namespace.